quesma/quesma/ui/ab_testing.go

```diff
@@ -0,0 +1,894 @@
+		buffer.Html("<thead>")
+		buffer.Html(`<tr>`)
+		buffer.Html(`<th class="key">Mismatch</th>`)
+		buffer.Html(`<th class="key">Count</th>`)
+		buffer.Html(`<th class="key"></th>`)
+		buffer.Html("</tr>")
+
+		buffer.Html("</thead>\n")
+		buffer.Html("<tbody>\n")
+
+		for _, row := range tableRows {
+			buffer.Html(`<tr>`)
+			buffer.Html(`<td>`)
+
+			mismatches, err := parseMismatches(row.mismatch)
+			if err == nil {
+				const limit = 10
+				size := len(mismatches)
+				if size > limit {
+					mismatches = mismatches[:limit]
+					mismatches = append(mismatches, jsondiff.JSONMismatch{
+						Message: fmt.Sprintf("... and %d more", size-limit),
+					})
+				}
+
+				buffer.Html(`<ol>`)
+				for _, m := range mismatches {
+					buffer.Html(`<li>`)
+					buffer.Html(`<p>`)
+					buffer.Text(m.Message)
+					buffer.Text(" ")
+
+					if m.Path != "" {
+						buffer.Html(`<code>`)
+						buffer.Text(`(`)
+						buffer.Text(m.Path)
+						buffer.Text(`)`)
+						buffer.Html(`</code>`)
+						{ // poor man's HTML indent
+							buffer.Html(`<ul>`)
+							buffer.Html(`<li>`)
+							buffer.Html(`<code>`)
+							buffer.Text("Actual: ")
+							buffer.Text(m.Actual)
+							buffer.Html(`</code>`)
+							buffer.Html(`</li>`)
+
+							buffer.Html(`<li>`)
+							buffer.Html(`<code>`)
+							buffer.Text("Expected: ")
+							buffer.Text(m.Expected)
+							buffer.Html(`</code>`)
+							buffer.Html(`</li>`)
+							buffer.Html(`</ul>`)
+						}
+					}
+					buffer.Html(`</p>`)
+					buffer.Html(`</li>`)
+				}
+				buffer.Html(`</ol>`)
+			} else {
+				buffer.Text(row.mismatch)
+			}
+			buffer.Html(`</td>`)
+
+			buffer.Html(`<td>`)
+			buffer.Text(fmt.Sprintf("%d", row.count))
+			buffer.Html(`</td>`)
+
+			buffer.Html("<td>")
+			buffer.Html(`<a href="`).Text(fmt.Sprintf("/ab-testing-dashboard/mismatch?dashboard_id=%s&panel_id=%s&mismatch_id=%s", dashboardId, panelId, row.mismatchId)).Html(`">`).Text("Requests").Html(`</a>`)
+			buffer.Html("</td>")
+
+			buffer.Html("</tr>\n")
+		}
+
+		buffer.Html("</tbody>\n")
+		buffer.Html("</table>\n")
+		buffer.Html("\n</main>\n\n")
+	} else {
+		buffer.Html(`<h3>No mismatches found</h3>`)
+
+	}
+
+	return buffer.Bytes()
+}
+
+func (qmc *QuesmaManagementConsole) generateABMismatchDetails(dashboardId, panelId, mismatchHash string) []byte {
+	buffer := newBufferWithHead()
+
+	dashboards, err := qmc.readKibanaDashboards()
+	dashboardName := dashboardId
+	panelName := panelId
+
+	if err == nil {
+		dashboardName = dashboards.dashboardName(dashboardId)
+		panelName = dashboards.panelName(dashboardId, panelId)
+	} else {
+		logger.Warn().Err(err).Msgf("Error reading dashboards %v", err)
+	}
+
+	buffer.Html(`<main id="ab_testing_dashboard">`)
+
+	buffer.Html(`<h2>A/B Testing - Panel requests</h2>`)
+
+	buffer.Html(`<h3>`)
+	buffer.Text(fmt.Sprintf("Dashboard: %s", dashboardName))
+	buffer.Html(`</h3>`)
+	buffer.Html(`<h3>`)
+	buffer.Text(fmt.Sprintf("Panel: %s", panelName))
+	buffer.Html(`</h3>`)
+
+	sql := `
```

Comment:
It would be good to extract this sql query execution to separate function

Reply:
I've separated these aspects in all handlers.